File: ios/AIRFacebook/Functions/InitFunction.m
/*
 * Copyright (c) 2018 Marcel Piestansky
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
 
#import "InitFunction.h"
#import "AIRFacebookEvent.h"
#import "AIRFacebook.h"
#import <AIRExtHelpers/MPFREObjectUtils.h>
#import "CachedAccessTokenLoader.h"
#import "ProfileTracker.h"
#import <FBSDKCoreKit/FBSDKCoreKit.h>
#import <AIRExtHelpers/MPUIApplicationDelegate.h>

FREObject fb_init( FREContext context, void *functionData, uint32_t argc, FREObject *argv ) {
    NSString* appID = [MPFREObjectUtils getNSString:argv[0]];
    NSString* urlSchemeSuffix = (argv[1] == nil) ? nil : [MPFREObjectUtils getNSString:argv[1]];
    BOOL showLogs = [MPFREObjectUtils getBOOL:argv[2]];
    int listenerID = [MPFREObjectUtils getInt:argv[3]];

    [[AIRFacebook sharedInstance] initWithAppID:appID urlSchemeSuffix:urlSchemeSuffix showLogs:showLogs];
    [AIRFacebook dispatchEvent:SDK_INIT];

    /* Start manual profile tracking */
    [ProfileTracker startTracking];
    /* Enable profile updates on access token change */
    [FBSDKProfile enableUpdatesOnAccessTokenChange:YES];
    /* Simulate didFinishLaunch notification so that a cached access token is loaded */
    [[FBSDKApplicationDelegate sharedInstance] application:[UIApplication sharedApplication] didFinishLaunchingWithOptions:[MPUIApplicationDelegate launchOptions]];
    /* Check cached access token */
    [CachedAccessTokenLoader checkAccessToken:listenerID];
    
    return nil;
}

File: android/src/com/marpies/ane/facebook/functions/InitFunction.java
/*
 * Copyright (c) 2018 Marcel Piestansky
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.marpies.ane.facebook.functions;

import com.adobe.fre.FREContext;
import com.adobe.fre.FREObject;
import com.facebook.FacebookSdk;
import com.marpies.ane.facebook.data.AIRFacebookEvent;
import com.marpies.ane.facebook.utils.AIR;
import com.marpies.ane.facebook.utils.AIRFacebookProfileTracker;
import com.marpies.ane.facebook.utils.CachedAccessTokenLoader;
import com.marpies.ane.facebook.utils.FREObjectUtils;

public class InitFunction extends BaseFunction implements FacebookSdk.InitializeCallback {

	@Override
	public FREObject call( FREContext freContext, FREObject[] args ) {
		super.call( freContext, args );

		if( FacebookSdk.isInitialized() ) return null;

		String appID = FREObjectUtils.getStringFromFREObject( args[0] );
		Boolean showLogs = FREObjectUtils.getBooleanFromFREObject( args[2] );
		int listenerID = FREObjectUtils.getIntFromFREObject( args[3] );
		AIR.setLogEnabled( showLogs );

		/* Call init with the Application ID */
		FacebookSdk.setApplicationId( appID );
		FacebookSdk.sdkInitialize( AIR.getContext().getActivity().getApplicationContext(), this );

		/* Start tracking profile changes */
		AIRFacebookProfileTracker.startTracking();
		/* Check cached access token */
		CachedAccessTokenLoader.load( listenerID );

		return null;
	}

	@Override
	public void onInitialized() {
		AIR.log( "Facebook SDK initialized | v" + FacebookSdk.getSdkVersion() );

		AIR.dispatchEvent( AIRFacebookEvent.SDK_INIT, "" );
	}

}
